examples/ota-requestor-app/esp32/main/OTARequesterImpl.cpp

```diff
@@ -0,0 +1,282 @@
+        return;
+    }
+
+    switch (*command)
+    {
+    case 1: { // QueryImage
+        chip::Callback::Cancelable * successCallback = mQueryImageResponseCallback.Cancel();
+        chip::Callback::Cancelable * failureCallback = mOnQueryImageFailureCallback.Cancel();
+
+        // These parameters are chosen arbitrarily
+        constexpr VendorId kExampleVendorId        = VendorId::Common;
+        constexpr uint16_t kExampleProductId       = CONFIG_DEVICE_PRODUCT_ID;
+        constexpr uint16_t kExampleHWVersion       = 0;
+        constexpr uint16_t kExampleSoftwareVersion = 0;
+        constexpr uint8_t kExampleProtocolsSupported =
+            EMBER_ZCL_OTA_DOWNLOAD_PROTOCOL_BDX_SYNCHRONOUS; // TODO: support this as a list once ember adds list support
+        const char locationBuf[] = { 'U', 'S' };
+        CharSpan exampleLocation(locationBuf);
+        constexpr bool kExampleClientCanConsent = false;
+        ByteSpan metadata;
+
+        err = cluster.QueryImage(successCallback, failureCallback, kExampleVendorId, kExampleProductId, kExampleHWVersion,
+                                 kExampleSoftwareVersion, kExampleProtocolsSupported, exampleLocation, kExampleClientCanConsent,
+                                 metadata);
+        if (err != CHIP_NO_ERROR)
+        {
+            ChipLogError(SoftwareUpdate, "QueryImage() failed: %s", chip::ErrorStr(err));
+        }
+        break;
+    }
+    case 2: { // ApplyUpdateRequest
+        chip::Callback::Cancelable * successCallback = mApplyUpdateResponseCallback.Cancel();
+        chip::Callback::Cancelable * failureCallback = mOnApplyUpdateRequestFailureCallback.Cancel();
+
+        constexpr uint32_t newVersion = 1;
+
+        err = cluster.ApplyUpdateRequest(successCallback, failureCallback, otaUpdateToken, newVersion);
+        if (err != CHIP_NO_ERROR)
+        {
+            ChipLogError(SoftwareUpdate, "ApplyUpdateRequest() failed: %s", chip::ErrorStr(err));
+        }
+        break;
+    }
+    case 3: { // NotifyUpdateApplied
+        constexpr uint32_t softwareVersion = 1;
+
+        err = cluster.NotifyUpdateApplied(nullptr, nullptr, otaUpdateToken, softwareVersion);
+        if (err != CHIP_NO_ERROR)
+        {
+            ChipLogError(SoftwareUpdate, "NotifyUpdateApplied() failed %s", chip::ErrorStr(err));
+        }
+        break;
+    }
+    default:
+        break;
+    }
+}
+
+void OnConnectionFailure(void * context, OperationalDeviceProxy * operationalDeviceProxy, CHIP_ERROR error)
+{
+    ChipLogError(SoftwareUpdate, "failed to connect to: %s", chip::ErrorStr(error));
+}
+
+void OnBlockReceived(void * context, chip::bdx::TransferSession::BlockData & blockdata)
+{
+    if (OTAUpdater::GetInstance().IsInProgress() == false)
+    {
+        OTAUpdater::GetInstance().Begin();
+    }
+    OTAUpdater::GetInstance().Write(reinterpret_cast<const void *>(blockdata.Data), blockdata.Length);
+}
```

Comment:
The data received will, at the beginning, include a Matter OTA Header, which may not be data that needs to be written (it's metadata). Examples right now in the SDK do not use the header, but it's in the spec. This will eventually need to handle the header processing/skipping before actually writing data.